bee-tangle/src/tangle.rs

```diff
@@ -1,759 +1,133 @@
-                entry.insert(vec![hash]);
-            }
-        }
-
-        if transaction.trunk() != transaction.branch() {
-            match self.approvers.entry(*transaction.branch()) {
-                Entry::Occupied(mut entry) => {
-                    let values = entry.get_mut();
-                    values.push(hash);
-                }
-                Entry::Vacant(entry) => {
-                    entry.insert(vec![hash]);
-                }
-            }
-        }
-
-        let vertex = Vertex::from(transaction, hash);
-
-        let tx_ref = vertex.get_ref_to_inner();
-
-        // TODO: not sure if we want replacement of vertices
-        if self.vertices.insert(hash, vertex).is_none() {
-            match self.solidifier_send.send(Some(hash)) {
-                Ok(()) => (),
-                Err(e) => todo!("log warning"),
-            }
-
-            Some(tx_ref)
-        } else {
-            None
+            children: DashMap::new(),
         }
     }
+}
 
-    pub(crate) fn shutdown(&self) {
-        // `None` will cause the worker to finish
-        self.solidifier_send.send(None).expect("error sending shutdown signal");
-        block_on(self.drop_barrier.wait());
-    }
-
-    /// Returns a reference to a transaction, if it's available in the local Tangle.
-    pub fn get_transaction(&'static self, hash: &Hash) -> Option<TransactionRef> {
-        self.vertices.get(hash).map(|v| v.get_ref_to_inner())
-    }
-
-    /// Returns whether the transaction is stored in the Tangle.
-    pub fn contains_transaction(&'static self, hash: &Hash) -> bool {
-        self.vertices.contains_key(hash)
+impl<T> Tangle<T>
+where
+    T: Clone + Copy,
+{
+    /// Creates a new Tangle.
+    pub fn new() -> Self {
+        Self::default()
     }
 
-    /// Returns whether the transaction associated with `hash` is solid.
-    ///
-    /// Note: This function is _eventually consistent_ - if `true` is returned, solidification has
-    /// definitely occurred. If `false` is returned, then solidification has probably not occurred,
-    /// or solidification information has not yet been fully propagated.
-    pub fn is_solid_transaction(&'static self, hash: &Hash) -> bool {
-        if self.is_solid_entry_point(hash) {
-            true
-        } else {
-            self.vertices.get(hash).map(|r| r.value().is_solid()).unwrap_or(false)
-        }
-    }
+    /// Inserts a transaction, and returns a thread-safe reference to it in case it didn't already exist.
+    pub fn insert(&self, data: Tx, hash: TxHash, metadata: T) -> Option<TxRef> {
+        self.add_child(*data.trunk(), hash);
 
-    /// Adds the `hash` of a milestone identified by its milestone `index`.
-    pub fn add_milestone(&'static self, index: MilestoneIndex, hash: Hash) {
-        self.milestones.insert(index, hash);
-        if let Some(mut vertex) = self.vertices.get_mut(&hash) {
-            vertex.set_milestone();
+        if data.trunk() != data.branch() {
```

Comment:
Just thinking, why not a set instead of a vec for the children ? We wouldn't have to worry about that.

Reply:
alright, makes sense, changed it.